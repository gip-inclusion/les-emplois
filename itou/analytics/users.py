from django.db.models import Count

from itou.users import enums as users_enums, models as users_models

from . import models


def collect_analytics_data(before):
    counts = {
        info["kind"]: info["count"]
        for info in users_models.User.objects.values("kind").annotate(count=Count("pk")).values("kind", "count")
    }
    return {
        models.DatumCode.USER_COUNT: sum(counts.values()),
        models.DatumCode.USER_JOB_SEEKER_COUNT: counts.get(users_enums.KIND_JOB_SEEKER, 0),
        models.DatumCode.USER_PRESCRIBER_COUNT: counts.get(users_enums.KIND_PRESCRIBER, 0),
        models.DatumCode.USER_EMPLOYER_COUNT: counts.get(users_enums.KIND_EMPLOYER, 0),
        models.DatumCode.USER_LABOR_INSPECTOR_COUNT: counts.get(users_enums.KIND_LABOR_INSPECTOR, 0),
        models.DatumCode.USER_ITOU_STAFF_COUNT: counts.get(users_enums.KIND_ITOU_STAFF, 0),
    }
