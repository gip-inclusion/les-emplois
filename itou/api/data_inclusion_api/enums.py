import enum


class StructureTypeStr(str, enum.Enum):
    ORGA = "orga"
    SIAE = "siae"
