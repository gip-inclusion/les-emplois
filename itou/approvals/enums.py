from django.db import models

from itou.companies import enums as companies_enums


class ApprovalStatus(models.TextChoices):
    EXPIRED = "EXPIRED", "Expiré"
    VALID = "VALID", "Valide"
    FUTURE = "FUTURE", "Valide (non démarré)"
    SUSPENDED = "SUSPENDED", "Valide (suspendu)"


class Origin(models.TextChoices):
    DEFAULT = "default", "Créé normalement via les emplois"
    PE_APPROVAL = "pe_approval", "Créé lors d'un import d'Agrément Pole Emploi"
    # On November 30th, 2021, AI were delivered approvals without a diagnosis.
    AI_STOCK = "ai_stock", "Créé lors de l'import du stock AI"
    ADMIN = "admin", "Créé depuis l'admin"


class ProlongationReason(models.TextChoices):
    SENIOR_CDI = "SENIOR_CDI", "CDI conclu avec une personne de plus de 57 ans"
    COMPLETE_TRAINING = "COMPLETE_TRAINING", "Fin d'une formation"
    RQTH = "RQTH", "RQTH - Reconnaissance de la qualité de travailleur handicapé"
    SENIOR = "SENIOR", "50 ans et plus"
    # Exclusive to AI and ACI
    PARTICULAR_DIFFICULTIES = (
        "PARTICULAR_DIFFICULTIES",
        "Difficultés particulièrement importantes dont l'absence de prise en charge ferait "
        "obstacle à son insertion professionnelle",
    )
    # Since December 1, 2021, health context reason can no longer be used
    HEALTH_CONTEXT = "HEALTH_CONTEXT", "Contexte sanitaire"

    @classmethod
    def for_company(cls, company):
        enums = [
            cls.SENIOR_CDI,
            cls.COMPLETE_TRAINING,
            cls.RQTH,
            cls.SENIOR,
        ]
        if company.kind in [companies_enums.CompanyKind.AI, companies_enums.CompanyKind.ACI]:
            enums.append(cls.PARTICULAR_DIFFICULTIES)

        empty = [(None, cls.__empty__)] if hasattr(cls, "__empty__") else []
        return empty + [(enum.value, enum.label) for enum in enums]


class ProlongationRequestStatus(models.TextChoices):
    PENDING = "PENDING", "À traiter"
    GRANTED = "GRANTED", "Acceptée"
    DENIED = "DENIED", "Refusée"


class ProlongationRequestDenyReason(models.TextChoices):
    IAE = "IAE", "L’IAE ne correspond plus aux besoins / à la situation de la personne."
    SIAE = "SIAE", "La typologie de SIAE ne correspond plus aux besoins / à la situation de la personne."
    DURATION = "DURATION", "La durée de prolongation demandée n’est pas adaptée à la situation du candidat."


class ProlongationRequestDenyProposedAction(models.TextChoices):
    # TODO: Clarify the actions to improve the naming
    EXIT_IAE = (
        "EXIT_IAE",
        "Accompagnement à la recherche d’emploi hors IAE et mobilisation de l’offre "
        "de services disponible au sein de votre structure ou celle d’un partenaire.",
    )
    SOCIAL_PARTNER = "SOCIAL_PARTNER", "Orientation vers un partenaire de l’accompagnement social/professionnel."
    OTHER = "OTHER", "Autre"
