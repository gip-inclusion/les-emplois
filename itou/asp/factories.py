import datetime
import random

import factory

from itou.asp import models


_sample_europe_countries = [
    {"code": "101", "name": "DANEMARK", "group": "2"},
    {"code": "111", "name": "BULGARIE", "group": "2"},
    {"code": "135", "name": "PAYS-BAS", "group": "2"},
]

_sample_outside_europe_countries = [
    {"code": "212", "name": "AFGHANISTAN", "group": "3"},
    {"code": "436", "name": "BAHAMAS", "group": "3"},
    {"code": "324", "name": "CONGO", "group": "3"},
]

_sample_france = [
    {"code": "714", "name": "BORA-BORA", "group": "1"},
    {"code": "812", "name": "KOUMAC", "group": "1"},
    {"code": "737", "name": "PUKAPUKA", "group": "1"},
]

_sample_communes = [
    {"code": "64483", "name": "SAINT-JEAN-DE-LUZ"},
    {"code": "97108", "name": "CAPESTERRE-DE-MARIE-GALANT"},
    {"code": "97107", "name": "CAPESTERRE-BELLE-EAU"},
    {"code": "37273", "name": "VILLE-AUX-DAMES"},
    {"code": "13200", "name": "MARSEILLE"},
    {"code": "67152", "name": "GEISPOLSHEIM"},
    {"code": "85146", "name": "MONTAIGU"},
]


class CountryFactory(factory.django.DjangoModelFactory):
    class Meta:
        model = models.Country
        django_get_or_create = ("code",)

    code, name, group = random.choice(_sample_europe_countries).values()


class CountryFranceFactory(CountryFactory):
    code, name, group = "100", "FRANCE", "1"


class CountryCollectiviteOutremerFactory(CountryFactory):
    """
    Parts or regions of France having their own INSEE country code

    Most of them are "Collectivit√©s d'Outre-Mer"
    """

    code, name, group = random.choice(_sample_france).values()


class CountryEuropeFactory(CountryFactory):
    code, name, group = random.choice(_sample_europe_countries).values()


class CountryOutsideEuropeFactory(CountryFactory):
    code, name, group = random.choice(_sample_outside_europe_countries).values()


class CommuneFactory(factory.django.DjangoModelFactory):
    """Factory for ASP INSEE commune:
    - if `code` or `name` are build parameters, object values will result of a lookup in `_sample_communes`
    - otherwise, fields `code` and `name` will be set from a randomly picked sample commune
    """

    class Meta:
        model = models.Commune
        django_get_or_create = (
            "code",
            "start_date",
            "end_date",
        )

    # FIXME: may cause issues in testing validity periods
    start_date = datetime.date(2000, 1, 1)
    end_date = None

    # Skipping definition of `code` and `name` fields
    # will be set in `_adjust_kwargs`

    @classmethod
    def _adjust_kwargs(cls, **kwargs):
        # Allow creation with parameters `code` or `name` (first matched)
        # either field must be a match in `_sample_communes`
        match kwargs:
            case {"code": code}:
                for item in _sample_communes:
                    if item["code"] == code:
                        kwargs["name"] = item["name"]
            case {"name": name}:
                for item in _sample_communes:
                    if item["name"] == name:
                        kwargs["code"] = item["code"]
            case _:
                code, name = random.choice(_sample_communes).values()
                kwargs["code"] = code
                kwargs["name"] = name

        return kwargs

    @classmethod
    def _adjust_kwargs(cls, **kwargs):
        # Allow creation with parameters `code` or `name` (first matched)
        # either field must be a match in `_sample_communes`
        match kwargs:
            case {"code": code}:
                for item in _sample_communes:
                    if item["code"] == code:
                        kwargs["name"] = item["name"]
            case {"name": name}:
                for item in _sample_communes:
                    if item["name"] == name:
                        kwargs["code"] = item["code"]
            case _:
                code, name = random.choice(_sample_communes).values()
                kwargs["code"] = code
                kwargs["name"] = name

        return kwargs


# FIXME: unreliable and confusing
class MockedCommuneFactory(CommuneFactory):
    """
    A factory with a specific code for mock testing
    """

    code, name = _sample_communes[-1].values()
