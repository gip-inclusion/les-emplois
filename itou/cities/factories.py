import json
import os

from django.contrib.gis.geos import GEOSGeometry, Point
from django.template.defaultfilters import slugify

from itou.cities.models import City


CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))

CITIES_JSON_FILE = f"{CURRENT_DIR}/management/commands/data/cities.json"


def create_test_cities(selected_departments, num_per_department=None):
    """
    Not a factory strictly speaking, but it's sitting here for discoverability.

    Pick cities from `cities.json` based on the given department numbers.
        create_test_cities(['01', '02'])

    The number of cities in each department can be limited:
        create_test_cities(['54', '57'], num_per_department=2)
    """
    departments_counter = {department: 0 for department in selected_departments}

    with open(CITIES_JSON_FILE, "r") as raw_json_data:

        json_data = json.load(raw_json_data)

        for item in json_data:

            department = item.get("codeDepartement")
            if (not department) or (department not in selected_departments):
                continue

            coords = item.get("centre")
            if not coords:
                continue

            departments_counter[department] += 1
            if num_per_department and departments_counter[department] > num_per_department:
                if all(value >= num_per_department for value in departments_counter.values()):
                    break
                continue

            City.objects.create(
                slug=slugify(f"{item['nom']}-{department}"),
                department=department,
                name=item["nom"],
                post_codes=item["codesPostaux"],
                code_insee=item["code"],
                coords=GEOSGeometry(f"{coords}"),  # Feed `GEOSGeometry` with GeoJSON.
            )


def create_city_saint_andre():
    return City.objects.create(
        name="Saint-André-des-Eaux",
        slug="saint-andre-des-eaux-44",
        department="44",
        coords=Point(-2.3140436, 47.3618584),
        post_codes=["44117"],
        code_insee="44117",
    )


def create_city_guerande():
    return City.objects.create(
        name="Guérande",
        slug="guerande-44",
        department="44",
        coords=Point(-2.4747713, 47.3358576),
        # Dummy
        post_codes=["44350"],
        code_insee="44350",
    )


def create_city_vannes():
    return City.objects.create(
        name="Vannes",
        slug="vannes-56",
        department="56",
        coords=Point(-2.8186843, 47.657641),
        # Dummy
        post_codes=["56000"],
        code_insee="56000",
    )
