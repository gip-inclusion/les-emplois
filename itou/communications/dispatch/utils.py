class JobSeekerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_job_seeker


class EmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_employer


class PrescriberNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and self.user.is_prescriber


class PrescriberOrEmployerNotification:
    def is_manageable_by_user(self):
        return super().is_manageable_by_user() and (self.user.is_prescriber or self.user.is_employer)
