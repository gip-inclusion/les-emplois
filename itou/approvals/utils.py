from itou.companies.enums import SIAE_WITH_CONVENTION_KINDS


def get_user_last_accepted_siae_job_application(user):
    if not user.is_job_seeker:
        return None

    # Some candidates may not have accepted job applications
    # Assuming it's the case can lead to issues downstream
    return (
        user.job_applications.accepted()
        .filter(to_company__kind__in=SIAE_WITH_CONVENTION_KINDS)
        .with_accepted_at()
        .order_by("-accepted_at", "-hiring_start_at")
        .first()
    )


def last_hire_was_made_by_siae(user, siae):
    if not user.is_job_seeker:
        return False
    last_accepted_job_application = get_user_last_accepted_siae_job_application(user)
    return last_accepted_job_application and last_accepted_job_application.to_company_id == siae.pk
