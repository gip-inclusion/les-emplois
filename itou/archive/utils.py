import datetime

from django.db import models
from django.db.models import Count, Exists, OuterRef, Q
from django.db.models.functions import Coalesce
from django.utils import timezone

from itou.approvals.models import Approval
from itou.eligibility.models import EligibilityDiagnosis, GEIQEligibilityDiagnosis
from itou.gps.models import FollowUpGroup
from itou.job_applications.models import JobApplication
from itou.users.models import User, UserKind


def get_filter_kwargs_on_user_for_related_objects_to_check():
    """
    Returns a dictionary of filter parameters to check for objects related
    to the User model that are not cascade deleted.
    """
    return {
        f"{obj.name}__isnull": True
        for obj in User._meta.related_objects
        if getattr(obj, "on_delete", None) and obj.on_delete != models.CASCADE
    }


def get_year_month_or_none(date=None):
    if not date:
        return None

    if isinstance(date, datetime.datetime):
        return timezone.localdate(date).replace(day=1)

    return date.replace(day=1)


def count_related_subquery(model, fk_field, outer_ref_field, extra_filters=None):
    filters = {fk_field: OuterRef(outer_ref_field)}
    if extra_filters:
        filters.update(extra_filters)
    return Coalesce(
        model.objects.filter(**filters).values(fk_field).annotate(count=Count(outer_ref_field)).values("count"), 0
    )


def inactive_jobseekers_without_recent_related_objects(inactive_since, notified, batch_size=None):
    recent_approval = Approval.objects.filter(
        Q(end_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), user_id=OuterRef("pk")
    )
    recent_eligibility_diagnosis = EligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_geiq_eligibility_diagnosis = GEIQEligibilityDiagnosis.objects.filter(
        Q(expires_at__gt=inactive_since) | Q(updated_at__gt=inactive_since), job_seeker_id=OuterRef("pk")
    )
    recent_job_application = JobApplication.objects.filter(job_seeker_id=OuterRef("pk"), updated_at__gt=inactive_since)
    recent_followup_group = FollowUpGroup.objects.filter(beneficiary_id=OuterRef("pk"), updated_at__gt=inactive_since)

    qs = (
        User.objects.filter(
            (Q(last_login__lte=inactive_since) | Q(last_login__isnull=True)),
            date_joined__lte=inactive_since,
            kind=UserKind.JOB_SEEKER,
            upcoming_deletion_notified_at__isnull=not notified,
        )
        .filter(
            ~Exists(recent_approval),
            ~Exists(recent_eligibility_diagnosis),
            ~Exists(recent_geiq_eligibility_diagnosis),
            ~Exists(recent_job_application),
            ~Exists(recent_followup_group),
        )
        .order_by("date_joined", "pk")
    )

    if batch_size:
        return qs[:batch_size]

    return qs
