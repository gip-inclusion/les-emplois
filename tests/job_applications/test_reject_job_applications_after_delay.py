import datetime

import pytest
from django.core.management import call_command
from django.utils import timezone

from itou.job_applications.enums import (
    AUTO_REJECT_JOB_APPLICATION_DELAY,
    AUTO_REJECT_JOB_APPLICATION_STATES,
    JobApplicationState,
    RefusalReason,
)
from tests.job_applications.factories import JobApplicationFactory
from tests.users.factories import JobSeekerFactory


@pytest.mark.parametrize("state", AUTO_REJECT_JOB_APPLICATION_STATES)
def test_reject_job_applications_after_delay(state, django_capture_on_commit_callbacks, mailoutbox, snapshot, caplog):
    limit = 2

    # first job_seeker with the oldest and the most recent job applications
    oldest_expected_job_application = JobApplicationFactory(
        state=state,
        answer="",
        updated_at=timezone.now() - AUTO_REJECT_JOB_APPLICATION_DELAY - datetime.timedelta(days=4),
    )
    recent_job_application_of_the_same_job_seeker = JobApplicationFactory(
        state=state,
        answer="",
        job_seeker=oldest_expected_job_application.job_seeker,
        updated_at=timezone.now() - AUTO_REJECT_JOB_APPLICATION_DELAY,
    )
    # second job_seeker with numerous oldish job applications
    job_seeker = JobSeekerFactory()
    other_expected_job_applications = JobApplicationFactory.create_batch(
        limit + 1,
        job_seeker=job_seeker,
        state=state,
        answer="",
        updated_at=timezone.now() - AUTO_REJECT_JOB_APPLICATION_DELAY - datetime.timedelta(days=3),
    )
    # unselected oldish job application of the third job_seeker
    unexpected_job_application = JobApplicationFactory(
        state=state,
        answer="",
        updated_at=timezone.now() - AUTO_REJECT_JOB_APPLICATION_DELAY - datetime.timedelta(days=2),
    )

    with django_capture_on_commit_callbacks(execute=True):
        call_command("reject_job_applications_after_delay", "--limit", limit)

    assert "5 auto rejected job applications for 2 job seekers." in caplog.messages

    # selected job applications
    for job_application, mail in list(
        zip(
            [
                oldest_expected_job_application,
                recent_job_application_of_the_same_job_seeker,
                *other_expected_job_applications,
            ],
            mailoutbox,
        )
    ):
        job_application.refresh_from_db()
        assert job_application.state == JobApplicationState.REFUSED
        assert job_application.refusal_reason == RefusalReason.AUTO
        assert job_application.answer == snapshot(name="auto.txt")
        assert mail.to == [job_application.job_seeker.email]
        assert job_application.answer in mail.body

    # unselected job application
    unexpected_job_application.refresh_from_db()
    assert unexpected_job_application.state == state
    assert unexpected_job_application.refusal_reason == ""
    assert unexpected_job_application.answer == ""
    assert unexpected_job_application.job_seeker.email not in [
        email_addr for mail in mailoutbox for email_addr in mail.to
    ]
