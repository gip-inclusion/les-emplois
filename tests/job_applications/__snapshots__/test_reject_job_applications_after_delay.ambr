# serializer version: 1
# name: test_reject_job_applications_after_delay[new][auto.txt]
  '''
  Votre candidature a été automatiquement déclinée car elle n’a pas reçu de réponse depuis plus de 2 mois.
  
  Si vous êtes toujours en recherche d’emploi, nous vous invitons à poursuivre vos démarches. Pour maximiser vos chances de retour à l’emploi, n’hésitez pas à vous faire accompagner par un prescripteur habilité (France Travail, Mission Locale, Cap emploi, Service social du Conseil départemental…).
  
  Pour trouver les prescripteurs habilités présents sur votre territoire, consultez notre moteur de recherche : https://emplois.inclusion.beta.gouv.fr/search/prescribers/results
  
  Nous vous souhaitons une pleine réussite dans vos démarches.
  
  '''
# ---
# name: test_reject_job_applications_after_delay[processing][auto.txt]
  '''
  Votre candidature a été automatiquement déclinée car elle n’a pas reçu de réponse depuis plus de 2 mois.
  
  Si vous êtes toujours en recherche d’emploi, nous vous invitons à poursuivre vos démarches. Pour maximiser vos chances de retour à l’emploi, n’hésitez pas à vous faire accompagner par un prescripteur habilité (France Travail, Mission Locale, Cap emploi, Service social du Conseil départemental…).
  
  Pour trouver les prescripteurs habilités présents sur votre territoire, consultez notre moteur de recherche : https://emplois.inclusion.beta.gouv.fr/search/prescribers/results
  
  Nous vous souhaitons une pleine réussite dans vos démarches.
  
  '''
# ---
